src/cleanup.ts

```diff
@@ -14,9 +19,59 @@ export async function run(cutoff = '24h'): Promise<void> {
     await docker.pruneNetworks({filters: untilFilter})
     core.info(`Pruning containers older than ${cutoff}`)
     await docker.pruneContainers({filters: untilFilter})
+    await cleanupOldImageVersions(docker, UPDATER_IMAGE_NAME)
+    await cleanupOldImageVersions(docker, PROXY_IMAGE_NAME)
   } catch (error) {
     core.error(`Error cleaning up: ${error.message}`)
   }
 }
 
+export async function cleanupOldImageVersions(
+  docker: Docker,
+  imageName: string
+): Promise<void> {
+  const repo = repositoryName(imageName)
+  const options = {
+    filters: {
+      reference: [repo]
+    }
+  }
+
+  core.info(`Cleaning up images for ${repo}`)
+
+  docker.listImages(options, async function (err, images) {
+    if (images && images.length > 0) {
+      for (const imageInfo of images) {
+        // The given imageName is expected to be a digest, however to avoid any surprises in future
+        // we fail over to check for a match on tags as well.
+        //
+        // This means we won't remove any image which matches an imageName of either of these notations:
+        // - dependabot/image@sha256:$REF (current implementation)
+        // - dependabot/image:v1
+        //
+        // Without checking imageInfo.RepoTags for a match, we would actually remove the latter even if
+        // this was the active version.
+        if (
+          imageInfo.RepoDigests?.includes(imageName) ||
+          imageInfo.RepoTags?.includes(imageName)
+        ) {
+          core.info(`Skipping current image ${imageInfo.Id}`)
+          continue
+        }
+
+        core.info(`Removing image ${imageInfo.Id}`)
+        try {
+          await docker.getImage(imageInfo.Id).remove()
+        } catch (error) {
+          if (error.statusCode === 409) {
```

Comment:
I had a stab at adding a unit test for this, but it proved to be a bit of a timesink as the GitHub-hosted image I'm using for the deletion tests is effectively blank with no binaries.

I tested this manually by basically opening a shell into an updater container and running `npm run cleanup-docker` and I figured this was good enough for now as I think the only way to more exhaustively test non-happy path code like this would be to incorporate a stand-in for the docker daemon, which I figured was out of scope for now.

Reply:
Should we log for other types of errors too? If there's something else that could go wrong here it would be tricky to debug without an indication this failed.

Reply:
That's a good call out, any other errors should bubble up to the `run()` method and get logged out there. I specifically rescued this to output it as `core.info` rather than `core.error` as it doesn't really need to be an annotation on the job since it is working as intended and should only be a transient state around upgrades.